### ISSUE TITLE
Example usage in package

### ISSUE THREAD

**Issue by kevinykuo** [mitre/mvndeps#1]
@mattpollock is there a repo that uses this package to manage dependencies?

**Comment by mattpollock** [comment]
None external to MITRE that I am aware of. We use it internally for several things. For example, we have lots of java models/parsers/etc with associated R packages. Instead of sticking the jars in the (internal) R repos, we define the java dependency (with version) that the package supports and use `mvndeps` to resolve dependencies in the `.onLoad()`.  After the first time, this amounts to double checking that the jars still exist in the `.m2` repo.

This works really nicely for it's intended purpose, but I haven't figured out how to "bring it to the masses" yet. The reason being the `mvndeps` assumes that users have maven installed and configured. This places a bit of a burden on R users who don't know any java.

**Comment by mattpollock** [comment]
Here is a sanitized version of a MITRE-internal use:

```r
#' @importFrom mvndeps get_classpath
#' @importFrom mvndeps get_dependency_jar
#' @importFrom mvndeps set_java_home
#' @importFrom mvndeps clear_java_home
#' @noRd
resolve_classpath <- function(pkgname) {
  
  # because it can take a little while for maven to do its thing
  message("Resolving java classpath using maven...")
  
  # trigger transitive dependency download if dep is not found
  deps <- c("org.mitre.caasd:java-dependency1:1.0.4", "com.google:java-dependency2:3.1.2")
  invisible(get_dependency_jar(dep=deps, quiet=TRUE))
  
  # get classpath
  cp <- get_classpath(deps, quiet=TRUE)

  # make sure all dependencies are available
  if (!all(file.exists(cp)))
    stop(paste("Required dependency", cp[!file.exists(cp)], 
               "not found. Resolve using mvndeps::download_dependency"))
  
  message("Java classpath resolved")
  
  return(cp)
}


#' @import rJava
.onLoad <- function(libname, pkgname) {  
 
  # set up java cp
  cp <- resolve_classpath(pkgname)
  .jpackage(name=pkgname, morePaths=cp)
}
```

**Comment by kevinykuo** [comment]
Do you think an `install_maven()` function or something similar would solve this?

Also, would the current design mean that the first time the user `library()` a package with dependencies, it could take a while since that's the time jars are resolved and downloaded?

**Comment by mattpollock** [comment]
Yes, and yes. Though The `resolve_classpath` method could be called from `.onAttach` or some other place. I don't know off the top of my head if `install.packages` triggers `.onAttach`.

It does take longer that it feels like it should (order of 5-10 seconds) for maven to check that dependencies exist even when they are already there. All the time in in the maven process though so I'm not sure there is anything to do about it.
